### ISSUE TITLE
Trouble compiling

### ISSUE THREAD

**Issue by JoaoMMR** [SCIInstitute/SCIRun#2092]
Hi everyone,

I've been having some problems compiling SCIRun and I was hoping someone could help me with this issue.
I start by running the CMake GUI and follow the instructions in the documentation but the building process is quite "fast". It doesn't appear to be compiling. In fact, no executable file is generated.

I've attached some images that show the project after configuration and the contents of the folder used for generation.
I'm using both the latest version of SCIRun and Windows 10

Thank you for considering my request.

![Image 1](https://user-images.githubusercontent.com/12515181/78141921-164cd480-7424-11ea-977e-395f50a623be.png)
![Image 2](https://user-images.githubusercontent.com/12515181/78141992-37152a00-7424-11ea-9e62-e16d0219626c.png)

**Comment by dcwhite** [comment]
My reply might've gone to the spam folder. Here's what it said: 

The generated file "Superbuild.sln" needs to be opened in Visual Studio 2017 (2019 works as well), and built there (preferably in Release mode). That part will take some time! What you've done so far is just the configure step in CMake. 

There are additional quirks related to building on Windows which I can help you with if you need it. However, the larger question is **why are you building SCIRun on Windows?** There is an installer available [here](https://github.com/SCIInstitute/SCIRun/releases), and developing in SCIRun is significantly easier on Mac or Linux.

**Comment by JoaoMMR** [comment]
Thanks for the reply. I'll start by opening the "Superbuild.sln" file in VS2017 (is it "better" to change to 2019?) and get back here with the results. 

To address the big question: The idea is to develop new SCIRun modules for the Materials Science field and, given that there's the assumption these users are more accustomed to conventional platforms and that they won't interact with the modules beyond the user experience, Windows looked like the best option. However, if you feel like ultimately changing to Linux is a better idea, I can do so.

Concerning the quirks of building in Windows, every help I can get is very much appreciated!

**Comment by dcwhite** [comment]
VS 2017 is what I use now, I just know that 2019 works via the Github Action virtual machine build; I haven't upgraded my tools yet. We're still using just C++11 features (maybe some 14), no C++17 just yet.

That's great you want to add new modules. I'd suggest whatever platform you're most comfortable developing on, use that--SCIRun can be developed on all of them. The overall process is definitely smoother from the command line on Mac or Linux though. 

And as a further guide, there is this manual for adding new modules: https://github.com/SCIInstitute/SCIRun/blob/master/src/Documentation/Manuals/SCIRun5ModuleGeneration/SCIRun5ModuleGeneration.pdf

**Comment by JoaoMMR** [comment]
I'll stick to VS 2017 then and begin by giving development on Windows a shot.

Thanks for the guide. Does it have any differences from:
https://sciinstitute.github.io/scirun.pages/SCIRun5ModuleGeneration.html ?

On a related note, should I use this particular thread to post my further inevitable questions about module generation on Windows? Or shoud I make a new thread?

PS: Still on the building process. By the way, I'm a bit confused as to what Qt version to use. The documentation says 5.12 or later so should I use 5.12.7?

**Comment by dcwhite** [comment]
I think that's the same version of that guide. The github pages site is undergoing some reorganization, so your link might not work soon.

To continue the thread, you can try our new StackOverflow team page. I just sent you an invite.

Latest Qt is best, 5.14 is what I'm using now.

**Comment by JoaoMMR** [comment]
Alright. Thanks for the invite and for the clarification

**Comment by JoaoMMR** [comment]
Ok. I did the following:

- Open the Superbuild.sln file in VS 2017
- Changed the mode from Debug to Release
- Clicked "Build Solution"

Doesn't seem like it worked though:
![image](https://user-images.githubusercontent.com/12515181/78271155-bded0400-7503-11ea-9060-57ce78e9c7fb.png)

And there's no executable file in the folder.
I've attached the full output as a txt.

What should be the next step?

[build output.txt](https://github.com/SCIInstitute/SCIRun/files/4422170/build.output.txt)

**Comment by dcwhite** [comment]
To build with Python into SCIRun on Windows, Visual Studio 2013 is still needed. You can download it for free here (https://visualstudio.microsoft.com/vs/older-downloads/). Install it and then the build should work. Alternatively, in the Superbuild cmake settings, turn off BUILD_WITH_PYTHON if you don't need that feature for now, then reconfigure/generate, and the build should be good. The executable will be in SCIRun/bin/SCIRun/[Release or Debug]. The last manual step is to copy some dlls into this folder, I'll send you that list when you get there.

**Comment by JoaoMMR** [comment]
a small question: I went with the first option and installed VS 2013. In this case, what Qt version should I use (given that, apparently, the newer versions of Qt only support VS 2015 and later) ?

**Comment by dcwhite** [comment]
Qt version does not change, as the Python module compilation does not depend on it. You don't even have to load 2013 during the build, it is called within VS2017's build of Superbuild.sln.

**Comment by JoaoMMR** [comment]
Tried it like this:

- Installed VS 2013
- Reran the CMake configuration with VS 12 2013 as the generator leaving the other options with their default settings (Win32, no optional toolset, use default native compilers). Left the Qt path like it was before (to VC 2017 x86). It definitely did more than before
- Open the Superbuild.sln file with VS 2013 and built the solution in release mode

Unfortunately it still didn't work. Got almost two thousand errors
[build output.txt](https://github.com/SCIInstitute/SCIRun/files/4430029/build.output.txt)

**Comment by dcwhite** [comment]
Sorry for the confusion (it's Windows' fault :). 

* You still use VS2017 to build SCIRun, and select that platform in CMake. The python-building Windows project files are hard-coded to only work with VS2013, which VS2017 will call automatically within the Superbuild solution. So nothing else needs to be done other than installing VS2013 to the standard location. Sorry about that.
* Also, a couple other quirks that came up since we have a student building on Windows for the first time here as well:
  * In newer CMake versions, you must explicitly set the platform to x64 (it defaults to Win32). SCIRun only supports 64-bit compilation. It's the second dropdown menu in the Configure window.
  * To get the Python build to work, in addition to VS2013, you need to configure the VS2017 install with a couple older Windows toolkit libraries, the exact versions of which I do not remember. When you get that far, I will figure that part out again and write it down :)

**Comment by dcwhite** [comment]
Yes, the build instruction page is in need of a refresh. This issue will be used to track updating http://sciinstitute.github.io/scirun.pages/build.html

**Comment by JoaoMMR** [comment]
I'm back to VS 2017 - now with VS 2013 as well. Everything seems to be fine in CMake (which didn't happen previously when only VS 2017 was installed).

Trying to build at the moment. What are the additional configurations to VS 2017?

**Comment by dcwhite** [comment]
Just make sure you're generating 64-bit projects. Older cmake looked like this, with 64 in the generator name:
![50018101-74a33900-ffce-11e8-990a-fe2211654e5c](https://user-images.githubusercontent.com/1399383/78575971-48c24d00-77e9-11ea-974f-9f04caf7be6b.png)
Newer cmake, you must select 64 in a separate dropdown menu (and VS2017 defaults to Win32, not 64):
![JCLyIta](https://user-images.githubusercontent.com/1399383/78576007-5546a580-77e9-11ea-97c9-ed9793501a72.png)

In both cases, there are no additional flags (I just grabbed those screenshots from the internet).

**Comment by JoaoMMR** [comment]
Well, the build finished and it looks like this still wasn't it. Still no executable file nor building without any failures (3 again).

This was the output:
[build output.txt](https://github.com/SCIInstitute/SCIRun/files/4439454/build.output.txt)

The CMake configuration I used:
![Capture](https://user-images.githubusercontent.com/12515181/78578266-2cdd9c00-7827-11ea-8a66-f77ea78d8900.PNG)

Just in case this is the issue, I'm pointing the Qt5 path to the msvc2017_64 folder.

**Comment by dcwhite** [comment]
So far so good. You've encountered the next two Windows build quirks.
* `18>CUSTOMBUILD : error : unable to create file libs/geometry/doc/html/geometry/reference/spatial_indexes/boost__geometry__index__rtree/rtree_parameters_type_const____indexable_getter_const____value_equal_const____allocator_type_const___.html: Filename too long
18>D	libs/geometry/doc/html/geometry/reference/spatial_indexes/boost__geometry__index__rtree/`
This needs to be solved by building in a "smaller" root directory, since Windows has a path character limit. I use the smallest possible, like `C:\_`, and that works. `C:\bin` should work too.
* For the python failure, I think it's the Windows SDK library version problem I mentioned above. Give me a minute and I'll paste the needed versions here.

**Comment by dcwhite** [comment]
Here is what else VS2017 needs to build that python library. Windows 8.1 SDK and multiple older Win10 SDKs:
<img width="348" alt="Screenshot 2020-04-06 10 27 48" src="https://user-images.githubusercontent.com/1399383/78581959-d3a74580-77f1-11ea-9831-4278a995da2d.png">
<img width="590" alt="Screenshot 2020-04-06 10 27 35" src="https://user-images.githubusercontent.com/1399383/78581962-d43fdc00-77f1-11ea-8070-5773a5536a45.png">

**Comment by dcwhite** [comment]
This might be quite frustrating, but it's good to finally have all these steps better-documented, for the next time. Thanks for being patient!

**Comment by JoaoMMR** [comment]
It is a bit but hey, at least I feel I'm contributing something to the community :)

Making the changes and attempting the build again

**Comment by JoaoMMR** [comment]
Still not working. Made the following changes:

- Changed the build folder to C:\_
- Downloaded and installed Windows 8.1 SDK and multiple older Win10 SDKs
![Capture](https://user-images.githubusercontent.com/12515181/78589271-daf14200-7837-11ea-9a2b-f37644f9b940.PNG)
![Capture1](https://user-images.githubusercontent.com/12515181/78589275-dcbb0580-7837-11ea-8e57-198d6ffa4d94.PNG)

The output:
[build output.txt](https://github.com/SCIInstitute/SCIRun/files/4441199/build.output.txt)

**Comment by dcwhite** [comment]
Hmm. Still the python project. It’s picking up the wrong header version. Did you reconfigure after installing the visual studio updates? Cmake might have cached the wrong library version or something. A reboot and clean build directory could fix it. Or there’s another step I need to think of...

**Comment by JoaoMMR** [comment]
By reconfiguring you mean the CMake configurations? Then no besides the one I've posted above. 
I'll try the reboot and clean build directory.

**Comment by JoaoMMR** [comment]
Got the same result

**Comment by dcwhite** [comment]
Well, at this point, I would normally walk over to the user's machine and fiddle with some stuff to get it to work. But since that is mostly impossible here/now, you have two main choices: 
1. Disable Python and start over, which would still enable virtually all of the legacy functionality in the available tutorials, as well as developing new modules in C++. The Python piece is entirely optional, but does of course add many layers of customization and programmability. However, even once built, on Windows it is still stuck back at version 3.4 and has many additional quirks to get it to do anything interesting (again, Mac and Linux work way better on that front). 
2. Switch to Mac or Linux or try the Ubuntu Subsystem for Linux inside Windows. I have built SCIRun w/Python and GUI on all three of those, but mainly develop on Mac and Windows now.

**Comment by dcwhite** [comment]
Alternatively, you could wait while I attempt to upgrade the Windows python version again (I ran into some different python build problems with newer python versions a while ago). Those might be ironed out in python 3.8, for instance.

**Comment by JoaoMMR** [comment]
Unfortunately, waiting is not an option. If anything I'm short on time.

1. Option 1 appears to detract from the whole thing. After all, the point is to develop Python modules though those "many additional quirks" you mentioned sound a bit scary. I suppose a solution could be to develop some sort of generic C++ module that would connect to external Python programs (or generate files that could be used by these programs) though I don't know how feasible nor desirable this is.

2. Option 2 is, for now, viewed as a last resort though I would prefer to attempt WSL before migrating to Linux.

Let me present a possible problem that, who knows, may be the cause of the ordeal:

The Python installation on this PC was done through the Anaconda package with all the other stuff it usually comes with. 
It looks to be living in its own bubble. For instance, I can't use Python through the usual command line, I have to use the anaconda prompt. 
Given that the problems I've been having with the build are due to Python, do you think removing Anaconda and reinstalling Python through either Windows installers or the Microsoft Store package could solve the issue?

**Comment by dcwhite** [comment]
Regarding (1), there is a SCIRun module that does exactly what you describe (named InterfaceWithPython), but it requires building Python within SCIRun to use, naturally. Currently there is no way to point SCIRun to a separate python installation as part of the build. Once SCIRun is built with its own python, it is possible to migrate python environments, even from Anaconda, to use other Python code. @jessdtate is the expert there, although he's mostly a Mac user. 

So, if you just want to develop SCIRun functionality from the Python side, even on Windows, compiling is not required--you can just install SCIRun using the installer, and Python is enabled and should work fine (with the usual python path/package quirks you'd might expect, seems you're familiar with those). This can get you up and running much faster and you can decide if the Python framework is sufficiently flexible. We're actively developing a new Python package that is being designed to work within SCIRun in a very general way, so any problem you run into ought to be worked on quickly.

**Comment by jessdtate** [comment]
It sounds like interfacewithpython will do what you need.  You can use one of the release builds for that.  Unfortunately, Windows is the hardest environment to work in.  You should be able to write some python that will work without any other setup.  Have a look at the python api docs (http://sciinstitute.github.io/scirun.pages/python.html) and post any questions on the email list or the forum.  

If you want to use existing packages that aren't in scirun's python (like numpy or scipy), that's where it gets a little tricky.  There are a few things that you can try.

**Comment by JoaoMMR** [comment]
It may but I don't known how to use it. I still have a lot of questions regarding development for SCIRun in general. 

@dcwhite you may want to check your inbox :)

**Comment by dcwhite** [comment]
Sorry, got a reply coming down the tubes shortly!

**Comment by dcwhite** [comment]
DRAFT STILL EDITING!
Additional steps for building with Python 3.8:

1. Code is on a branch now, will soon be in master. See PR #
1. Start a fresh build. Default Python version on Windows will be 3.8.
1. Once Superbuild is generated, start building it. It will fail at some point.
1. The problem now should be that the Python windows build is a 2-step process, and Superbuild does not know about that yet. So go into `bin/Externals/Source/Python_external/PCbuild`, and open a Powershell (command prompt might work too). Run `build.bat`.
1. With previous versions, this script would fail for various reasons. Hopefully, now with 3.8, it should finish successfully.
1. Once the script is done, return to Superbuild.sln, click build all, and the build should be successful. 
1. Executable is in bin/SCIRun/Release. But wait, there's one more quirk (and this still applies to my builds too):
1. There's a missing CMake step that should copy some dlls into the bin directory. Here's what you need to copy manually:

**Comment by dcwhite** [comment]
This just got easier: once the latest code is merged, here are the two commands that will build Windows with Python (replacing steps 2-6 above, but not 7-10) (adjust your Qt path accordingly):
```
cd SCIRun\bin
"C:\Program Files\CMake\bin\cmake" ..\Superbuild -DCMAKE_BUILD_TYPE=Release -G"Visual Studio 15 2017" -Ax64 -DQt5_PATH:PATH="C:/Qt5.14/5.14.0/msvc2017_64"
"C:\Program Files\CMake\bin\cmake" --build . --config Release -j6
```

**Comment by dcwhite** [comment]
I'll turn this into a script that will work just like the current `build.bat` for Mac/Linux.

**Comment by JoaoMMR** [comment]
On it. The branch is https://github.com/SCIInstitute/SCIRun/tree/programmable-port-prototype right?

**Comment by dcwhite** [comment]
Yup, that’s the one! Make sure to start fresh so the newer Python is downloaded.

**Comment by JoaoMMR** [comment]
Oh, one thing I didn't undestand then. 
Sorry if this question is dumb, do I need to download Python 3.8 first?

**Comment by dcwhite** [comment]
No, it’s part of the build process.

**Comment by JoaoMMR** [comment]
Steps:

No problems with steps 1 and 2.

3. There was no problem per se but it looks very much like the results I got before. Also, the generation process was quite quick - unlike the more "this is clearly doing something" I had with VS 2017 + 2013.

4. This may also not be an error (I am quite inexperienced with Powershell) but it seems like nothing happened. I had no results from running the last command. 
![Capture](https://user-images.githubusercontent.com/12515181/79492597-1ea63180-8018-11ea-9ad2-bb790665d12a.PNG)

**Comment by dcwhite** [comment]
You might need to install certain Visual Studio command-line tools, let me look into that one.

**Comment by JoaoMMR** [comment]
any progress?

**Comment by dcwhite** [comment]
I've been ironing out the (hopefully) last build issues with the python upgrade. Have you tried the new script on the PR branch? The syntax is `windowsBuild.bat <path-to-Qt-for-cmake>`. It's not as full-featured as `build.sh`, but ought to do what you need. For msbuild, you might just need to add it to the global windows path variable.

**Comment by dcwhite** [comment]
https://github.com/SCIInstitute/SCIRun/pull/2103

**Comment by JoaoMMR** [comment]
Not yet but I'll try it. Is the script intended to do those steps you mentioned previously?

**Comment by dcwhite** [comment]
Yes, it ought to automate everything.

**Comment by JoaoMMR** [comment]
I haven't found `windowsBuild.bat` but I did find `makeWindowsInstaller.bat`
I ran it (without specifying `<path-to-Qt-for-cmake>`) and no errors popped up but I'm not seeing any results - to some extent. A new folder named `installerbuild_` was made inside `scripts` which appears to be a copy of the entire SCIRun folder.

![1](https://user-images.githubusercontent.com/12515181/80981221-94d0d380-8e21-11ea-8123-009072a97682.PNG)
![2](https://user-images.githubusercontent.com/12515181/80981229-969a9700-8e21-11ea-9820-871b7733521c.PNG)

Am I supposed to follow the previous steps? So should I now go back to build with CMake?
If so, what Superbuild should I use? The new one or the original?

**Comment by JoaoMMR** [comment]
sorry about closing. That was a misclick

**Comment by dcwhite** [comment]
It's in the root directory, next to build.sh: https://github.com/SCIInstitute/SCIRun/pull/2103/files#diff-261649f7dedc8409416059a6b85c2f37. But it's not merged into master yet (our PR merge process is slow nowadays), you have to be on the `programmable-port-prototype` branch.

**Comment by JoaoMMR** [comment]
Whoops. My bad. 

Downloaded the correct branch and ran the bat file as instructed. Used msvc2017_64.
This was the result:
[build.txt](https://github.com/SCIInstitute/SCIRun/files/4581997/build.txt)

**Comment by dcwhite** [comment]
It's the same error as earlier, too long file name:  Filename too long [C:\Users\king_\Desktop\SCIRun\bin\Boost_external.vcxproj]. Try cloning in C:\<very short directory like "dev">\SCIRun\ and run the script from there, that worked on my machine. 

I'll add a warning to the script for this though! And also make it easier to build in a different directory, to avoid having to reclone.

**Comment by JoaoMMR** [comment]
on it

**Comment by JoaoMMR** [comment]
The tip worked. As of this moment, it is has been about roughly 3 hours since execution started. 
Is this normal by the way? This has taken considerably longer than any other previous attempt.

**Comment by JoaoMMR** [comment]
The process ended - I think - successfully. There were plenty of "No install step for x" and warnings but no errors and, finally, the SCIRun.exe

**Comment by dcwhite** [comment]
Hooray! And it only took 5 weeks ;) At least we got a decent script out of all this back and forth, and newer python to boot.

**Comment by JoaoMMR** [comment]
Exactly. Remember diminishing returns? This is like increasing returns. Just imagine what we could've achieved if it took 6 weeks :) 

Ok! Last questions before we close this thread:

1) Still concerning this build vs. the latest binary available, are there any major differences besides the Python version? 
For example, I always get an error with the old version about creating a DB session:
![image](https://user-images.githubusercontent.com/12515181/81196747-566d1d00-8fb7-11ea-8524-153b5fcaa79b.png)
The error is absent in this build:
![Capture](https://user-images.githubusercontent.com/12515181/81197138-d5faec00-8fb7-11ea-9aea-9be6dcded29e.PNG)

2) This should be an easy one. I have several problems with SCIRun itself which I think have mostly to do with the GUI. I should create a new thread right?

**Comment by dcwhite** [comment]
New month, time to catch up. I replied to your old email. Here are these answers:
1. That sqlite warning I haven't seen, but it's part of a separate feature that never got off the ground, and can be ignored either way--it shouldn't affect any Python functionality.
1. And yes, more new issues are always better than adding onto old ones. I'll close this one now, and make a new one to improve the windows script.
